pkg/container/cri/types.go

```diff
@@ -0,0 +1,73 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cri
+
+/*
+These types are from
+https://github.com/kubernetes/kubernetes/blob/063e7ff358fdc8b0916e6f39beedc0d025734cb1/pkg/kubelet/apis/cri/runtime/v1alpha2/api.pb.go#L183
+*/
+
+// Mount specifies a host volume to mount into a container.
+// This is a copy of the upstream cri Mount type
+// see: k8s.io/kubernetes/pkg/kubelet/apis/cri/runtime/v1alpha2
+// It additionally serializes the "propogation" field with the string enum
+// names on disk as opposed to the int32 values
+// In yaml this looks like:
+//  container_path: /foo
+//  host_path: /bar
+//  readonly: true
+//  selinux_relabel: false
+//  propagation: PROPAGATION_PRIVATE
+// Propogation may be one of:
+// PROPAGATION_PRIVATE, PROPAGATION_HOST_TO_CONTAINER, PROPAGATION_BIDIRECTIONAL
+type Mount struct {
+	// Path of the mount within the container.
+	ContainerPath string `protobuf:"bytes,1,opt,name=container_path,json=containerPath,proto3" json:"container_path,omitempty"`
+	// Path of the mount on the host. If the hostPath doesn't exist, then runtimes
+	// should report error. If the hostpath is a symbolic link, runtimes should
+	// follow the symlink and mount the real destination to container.
+	HostPath string `protobuf:"bytes,2,opt,name=host_path,json=hostPath,proto3" json:"host_path,omitempty"`
+	// If set, the mount is read-only.
+	Readonly bool `protobuf:"varint,3,opt,name=readonly,proto3" json:"readonly,omitempty"`
+	// If set, the mount needs SELinux relabeling.
+	SelinuxRelabel bool `protobuf:"varint,4,opt,name=selinux_relabel,json=selinuxRelabel,proto3" json:"selinux_relabel,omitempty"`
+	// Requested propagation mode.
+	Propagation MountPropagation `protobuf:"varint,5,opt,name=propagation,proto3,enum=runtime.v1alpha2.MountPropagation" json:"propagation,omitempty"`
```

Comment:
looks like these variable names need "adjustments" upstream.

Reply:
Yeah I'm not sure if that's actually an option. I think it would probably be fine for us to match the types closely but use a different on-disk format.

Reply:
we still have the option to shim with getters and setters. this means some extra boiler plate.
this fixes both the field names and hides the upstream config.

> create a CRI package, with types (Mount related only for now) copied from upstream
* this allows us to avoid stability concerns, if necessary we can provide our own conversion

i need to know the roadmap for CRI in terms of do they have plans to remove or rename fields, because if there are upstream API changes our local type can end up invalid and we will have to break kind users because these fields are public in the kind config.

i will leave this to you at this point if you want us to proceed, but comments from others would be appreciated too.

Reply:
CRI is pretty stable at this point, this type hasn't actually changed in years.

But our local type can be a shim just closely matching their type, if they diverge we will translate. As long as we are not using the client this isn't an issue.

But also for the types we expose here for on disk purposes, what they are capable of is unlikely to change. They cannot remove support for these fields without breaking Kubernetes majorly.

Reply:
ok

Reply:
It is nice for now that they are literally the same definition, but if it diverges slightly (EG they add a field or break the type up) we can keep our changes compatible and translate.

The things we will need to expose in config like Mounts, DNS, ... those seem unlikely to change significantly.

More complex types like entire containers, those seem stable too but we don't need to expose those on disk anyhow.

Reply:
I adjusted our names, strangely the proto json tags were already camelCase but not the actual json tags 🤷‍♂️ 

Where possible these match user facing API fields in Kubernetes APIs